src/couch_views/src/couch_views_http.erl

```diff
@@ -0,0 +1,272 @@
+    {_N, Results} = lists:foldl(fun(Args0, {Limit, Acc}) ->
+        case Limit > 0 of
+            true ->
+                Args = set_limit(Args0#mrargs{page_size = Limit}),
+                {Meta, Items} = Fun(Args),
+                Result = maybe_add_bookmark(
+                    PageSize, Args, Meta, Items, KeyFun),
+                #{total_rows := Total} = Result,
+                {Limit - Total, [Result | Acc]};
+            false ->
+                Bookmark = bookmark_encode(Args0),
+                Result = #{
+                    rows => [],
+                    next => Bookmark,
+                    total_rows => 0
+                },
+                {Limit, [Result | Acc]}
+        end
+    end, {PageSize, []}, QueriesArgs),
+    lists:reverse(Results).
+
+
+maybe_add_bookmark(PageSize, Args0, Response, Items, KeyFun) ->
+    #mrargs{page_size = Limit} = Args0,
+    Args = Args0#mrargs{page_size = PageSize},
+    case check_completion(Limit, Items) of
+        {Rows, nil} ->
+            maps:merge(Response, #{
+                rows => Rows,
+                total_rows => length(Rows)
+            });
+        {Rows, Next} ->
+            NextKey = KeyFun(Next),
+            if is_binary(NextKey) -> ok; true ->
+                throw("Provided KeyFun should return binary")
+            end,
+            Bookmark = bookmark_encode(Args#mrargs{start_key=NextKey}),
+            maps:merge(Response, #{
+                rows => Rows,
+                next => Bookmark,
+                total_rows => length(Rows)
+            })
+    end.
+
+
+set_limit(#mrargs{page_size = PageSize, limit = Limit} = Args)
+        when is_integer(PageSize) andalso Limit > PageSize ->
+    Args#mrargs{limit = PageSize + 1};
+
+set_limit(#mrargs{page_size = PageSize, limit = Limit} = Args)
+        when is_integer(PageSize)  ->
+    Args#mrargs{limit = Limit + 1}.
+
+
+check_completion(Limit, Items) when length(Items) > Limit ->
+    case lists:split(Limit, Items) of
+        {Head, [NextItem | _]} ->
+            {Head, NextItem};
+        {Head, []} ->
+            {Head, nil}
+    end;
+
+check_completion(_Limit, Items) ->
+    {Items, nil}.
+
+
+bookmark_encode(Args0) ->
+    N = ((record_info(size, mrargs) - 1) div 8 + 1) bsl 3,
+    {Term, Mask, _} = lists:foldl(fun(Value, {Acc, <<MI:N>> = M, Idx}) ->
+        case element(Idx, #mrargs{}) of
+            Value ->
+                {Acc, M, Idx + 1};
+            _Default when Idx == #mrargs.bookmark ->
+                {Acc, M, Idx + 1};
+            _Default ->
+                {[Value | Acc], <<((1 bsl Idx) bor MI):N>>, Idx + 1}
+        end
+    end, {[], <<0:N>>, 1}, tuple_to_list(Args0)),
+    Binary = term_to_binary(lists:reverse(Term), [compressed, {minor_version, 2}]),
+    couch_util:encodeBase64Url(<<?BOOKMARK_VSN, Mask/binary, Binary/binary>>).
+
+
+bookmark_decode(Bookmark) ->
+    try
+        N = ((record_info(size, mrargs) - 1) div 8 + 1) bsl 3,
```

Comment:
Should we just harcode the size of the mrargs and rely on CI to catch the case when mrargs is updated but mask size is not?

Reply:
I miss compile time computations in Erlang a lot.